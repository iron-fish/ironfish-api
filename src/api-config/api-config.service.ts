/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

@Injectable()
export class ApiConfigService {
  constructor(private readonly config: ConfigService) {}

  get<T = unknown>(name: string): T {
    const value = this.config.get<T>(name);
    if (value === undefined) {
      throw new Error(`Missing config value for ${name}`);
    }
    return value;
  }

  getWithDefault<T = unknown>(name: string, defaultValue: T): T {
    const value = this.config.get<T>(name);
    if (value === undefined) {
      return defaultValue;
    }
    return value;
  }

  isLocal(): boolean {
    return this.get<string>('NODE_ENV') === 'development';
  }

  isStaging(): boolean {
    return this.get<string>('NODE_ENV') === 'staging';
  }

  isProduction(): boolean {
    return this.get<string>('NODE_ENV') === 'production';
  }

  get dbPoolUrl(): string {
    return `${this.get<string>('DATABASE_CONNECTION_POOL_URL')}?pgbouncer=true`;
  }
}
