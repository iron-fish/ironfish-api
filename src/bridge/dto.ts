/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

export type IdRetrievalDTO = { [key: number]: string | null };

export type AddressCreationDTO = { [key: string]: number };

export type HeadHash = { hash: string };

export type OptionalHeadHash = { hash: string | null };
