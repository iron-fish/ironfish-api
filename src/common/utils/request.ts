/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Request } from 'express';

export function fetchIpAddressFromRequest(request: Request): string {
  const xForwardedFor = request.header('X-Forwarded-For');
  if (xForwardedFor) {
    const addresses = xForwardedFor.split(',');
    if (addresses.length) {
      return addresses[0].trim();
    }
  }

  // In Heroku, `X-Forwarded-For` will always be set. However, this will be used
  // as a fallback in local development.
  return request.ip;
}
