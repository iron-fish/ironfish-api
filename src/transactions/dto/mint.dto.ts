/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { IsString } from 'class-validator';

export class MintDto {
  @IsString()
  readonly id!: string;

  @IsString()
  readonly metadata!: string;

  @IsString()
  readonly name!: string;

  @IsString()
  readonly owner!: string;

  @IsString()
  readonly value!: string;
}
