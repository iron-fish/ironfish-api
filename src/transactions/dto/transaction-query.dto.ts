/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import { IsBoolean, IsOptional, IsString } from 'class-validator';

export class TransactionQueryDto {
  @ApiProperty({ description: 'Transaction hash' })
  @IsString()
  readonly hash!: string;

  @ApiPropertyOptional({
    description: 'Whether or not to include blocks in the response',
  })
  @IsOptional()
  @IsBoolean()
  @Type(() => Boolean)
  readonly with_blocks?: boolean;
}
