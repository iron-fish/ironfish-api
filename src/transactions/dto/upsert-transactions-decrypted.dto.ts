/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { Type } from 'class-transformer';
import {
  ArrayMaxSize,
  ArrayMinSize,
  IsArray,
  IsInt,
  IsString,
  ValidateNested,
} from 'class-validator';

export class NoteDecryptedDto {
  @IsInt()
  @Type(() => Number)
  readonly amount!: number;

  @IsString()
  readonly memo!: string;
}

export class TransactionDecryptedDto {
  @IsString()
  readonly hash!: string;

  @IsArray()
  @ValidateNested({ each: true })
  readonly notes!: NoteDecryptedDto[];
}

export class UpsertTransactionsDecryptedDto {
  @IsArray()
  @ArrayMinSize(1)
  @ArrayMaxSize(3000)
  @ValidateNested({ each: true })
  @Type(() => TransactionDecryptedDto)
  readonly transactions!: TransactionDecryptedDto[];
}
