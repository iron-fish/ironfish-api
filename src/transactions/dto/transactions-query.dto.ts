/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Type } from 'class-transformer';
import { IsBoolean, IsOptional, IsString } from 'class-validator';
import { PaginationArgsDto } from '../../common/dto/pagination-args.dto';

export class TransactionsQueryDto extends PaginationArgsDto {
  @IsOptional()
  @IsString()
  readonly search?: string;

  @IsOptional()
  @IsBoolean()
  @Type(() => Boolean)
  readonly with_block?: boolean;
}
