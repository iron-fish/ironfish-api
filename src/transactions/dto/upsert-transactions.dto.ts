/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { Type } from 'class-transformer';
import {
  ArrayMaxSize,
  ArrayMinSize,
  IsArray,
  IsInt,
  IsString,
  Max,
  ValidateNested,
} from 'class-validator';
import { NoteDto } from './note.dto';
import { SpendDto } from './spend.dto';

export class TransactionDto {
  @IsString()
  readonly hash!: string;

  @Max(Number.MAX_SAFE_INTEGER)
  @IsInt()
  @Type(() => Number)
  readonly fee!: number;

  @Max(Number.MAX_SAFE_INTEGER)
  @IsInt()
  @Type(() => Number)
  readonly size!: number;

  @IsArray()
  @ValidateNested({ each: true })
  readonly notes!: NoteDto[];

  @IsArray()
  @ValidateNested({ each: true })
  readonly spends!: SpendDto[];
}

export class UpsertTransactionsDto {
  @IsArray()
  @ArrayMinSize(1)
  @ArrayMaxSize(3000)
  @ValidateNested({ each: true })
  @Type(() => TransactionDto)
  readonly transactions!: TransactionDto[];
}
