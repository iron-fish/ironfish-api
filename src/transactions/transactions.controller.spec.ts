/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { HttpStatus, INestApplication } from '@nestjs/common';
import faker from 'faker';
import request from 'supertest';
import { v4 as uuid } from 'uuid';
import { ApiConfigService } from '../api-config/api-config.service';
import { PrismaService } from '../prisma/prisma.service';
import { bootstrapTestApp } from '../test/test-app';
import { UpsertTransactionsDto } from './dto/upsert-transactions.dto';
import { Transaction } from '.prisma/client';

const API_KEY = 'test';

describe('TransactionsController', () => {
  let app: INestApplication;
  let config: ApiConfigService;
  let prisma: PrismaService;

  beforeAll(async () => {
    app = await bootstrapTestApp();
    config = app.get(ApiConfigService);
    prisma = app.get(PrismaService);
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  const seedBlock = async () => {
    const hash = uuid();
    const sequence = faker.datatype.number();
    const searchable_text = hash + ' ' + String(sequence);

    const block = await prisma.block.create({
      data: {
        hash,
        difficulty: faker.datatype.number(),
        main: true,
        sequence,
        timestamp: new Date(),
        transactions_count: 0,
        graffiti: uuid(),
        previous_block_hash: uuid(),
        network_version: 0,
        searchable_text,
        size: faker.datatype.number(),
      },
    });

    return { block };
  };

  describe('POST /transactions', () => {
    beforeEach(() => {
      jest.spyOn(config, 'get').mockImplementationOnce(() => API_KEY);
    });

    describe('with a missing API key', () => {
      it('returns a 401', async () => {
        const { body } = await request(app.getHttpServer())
          .post(`/transactions`)
          .expect(HttpStatus.UNAUTHORIZED);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with missing arguments', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .post(`/transactions`)
          .set('Authorization', `Bearer ${API_KEY}`)
          .send({ transactions: [{}] })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with too many transactions', () => {
      it('returns a 422', async () => {
        const transactions = [];
        const { block } = await seedBlock();
        const notes = [{ commitment: uuid() }];
        const spends = [{ nullifier: uuid() }];

        for (let i = 0; i < 3001; i++) {
          transactions.push({
            hash: uuid(),
            fee: faker.datatype.number(),
            size: faker.datatype.number(),
            timestamp: new Date(),
            block_id: block.id,
            notes,
            spends,
          });
        }

        const payload: UpsertTransactionsDto = { transactions };

        const { body } = await request(app.getHttpServer())
          .post(`/transactions`)
          .set('Authorization', `Bearer ${API_KEY}`)
          .send(payload)
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a valid payload', () => {
      it('upserts transactions', async () => {
        const { block } = await seedBlock();
        const notes = [{ commitment: uuid() }];
        const spends = [{ nullifier: uuid() }];
        const payload: UpsertTransactionsDto = {
          transactions: [
            {
              hash: uuid(),
              fee: faker.datatype.number(),
              size: faker.datatype.number(),
              timestamp: new Date(),
              block_id: block.id,
              notes,
              spends,
            },
          ],
        };

        const transaction = payload.transactions[0];
        const { body } = await request(app.getHttpServer())
          .post(`/transactions`)
          .set('Authorization', `Bearer ${API_KEY}`)
          .send(payload)
          .expect(HttpStatus.CREATED);

        const { data } = body;
        expect((data as unknown[]).length).toBeGreaterThan(0);
        expect((data as unknown[])[0]).toMatchObject({
          id: expect.any(Number),
          hash: transaction.hash,
          fee: transaction.fee.toString(),
          size: transaction.size,
          timestamp: transaction.timestamp.toISOString(),
          block_id: transaction.block_id,
          notes: notes,
          spends: spends,
        });
      });
    });
  });

  describe('GET /transactions/find', () => {
    describe('with a valid hash', () => {
      it('returns the transaction with the correct hash', async () => {
        const testTransactionHash = uuid();
        const { block } = await seedBlock();
        const notes = [{ commitment: uuid() }];
        const spends = [{ nullifier: uuid() }];
        await prisma.transaction.create({
          data: {
            hash: testTransactionHash,
            network_version: 0,
            fee: faker.datatype.number(),
            size: faker.datatype.number(),
            timestamp: new Date(),
            block_id: block.id,
            notes,
            spends,
          },
        });

        const { body } = await request(app.getHttpServer())
          .get('/transactions/find')
          .query({ hash: testTransactionHash })
          .expect(HttpStatus.OK);

        expect(body).toMatchObject({
          id: expect.any(Number),
          hash: testTransactionHash,
          fee: expect.any(String),
          size: expect.any(Number),
          timestamp: expect.any(String),
          block_id: block.id,
          notes: notes,
          spends: spends,
        });
      });
    });

    describe('with an invalid hash', () => {
      it('returns a 404', async () => {
        const { block } = await seedBlock();
        const notes = [{ commitment: uuid() }];
        const spends = [{ nullifier: uuid() }];
        await prisma.transaction.create({
          data: {
            hash: uuid(),
            network_version: 0,
            fee: faker.datatype.number(),
            size: faker.datatype.number(),
            timestamp: new Date(),
            block_id: block.id,
            notes,
            spends,
          },
        });

        const { body } = await request(app.getHttpServer())
          .get('/transactions/find')
          .query({ hash: uuid() })
          .expect(HttpStatus.NOT_FOUND);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with an undefined hash', () => {
      it('returns a 422', async () => {
        const { block } = await seedBlock();
        const notes = [{ commitment: uuid() }];
        const spends = [{ nullifier: uuid() }];
        await prisma.transaction.create({
          data: {
            hash: uuid(),
            network_version: 0,
            fee: faker.datatype.number(),
            size: faker.datatype.number(),
            timestamp: new Date(),
            block_id: block.id,
            notes,
            spends,
          },
        });

        const { body } = await request(app.getHttpServer())
          .get('/transactions/find')
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });
  });

  describe('GET /transactions', () => {
    describe('with a valid partial hash search string', () => {
      it('retuns transactions with match(es)', async () => {
        const { block } = await seedBlock();
        const testTransactionHash = uuid();
        const notes = [{ commitment: uuid() }];
        const spends = [{ nullifier: uuid() }];
        const transaction = await prisma.transaction.create({
          data: {
            hash: testTransactionHash,
            network_version: 0,
            fee: faker.datatype.number(),
            size: faker.datatype.number(),
            timestamp: new Date(),
            block_id: block.id,
            notes,
            spends,
          },
        });

        const { body } = await request(app.getHttpServer())
          .get('/transactions')
          .query({ search: testTransactionHash.slice(0, 5) })
          .expect(HttpStatus.OK);

        const { data } = body;
        expect((data as unknown[]).length).toBeGreaterThan(0);
        expect((data as unknown[])[0]).toMatchObject({
          id: expect.any(Number),
          hash: transaction.hash,
          fee: transaction.fee.toString(),
          size: transaction.size,
          timestamp: transaction.timestamp.toISOString(),
          block_id: transaction.block_id,
          notes: notes,
          spends: spends,
        });
      });
    });

    describe('with no query parameters', () => {
      it('retuns transactions with match(es)', async () => {
        const { block } = await seedBlock();
        const notes = [{ commitment: uuid() }];
        const spends = [{ nullifier: uuid() }];
        for (let i = 0; i < 10; i++) {
          await prisma.transaction.create({
            data: {
              hash: uuid(),
              network_version: 0,
              fee: faker.datatype.number(),
              size: faker.datatype.number(),
              timestamp: new Date(),
              block_id: block.id,
              notes,
              spends,
            },
          });
        }

        const { body } = await request(app.getHttpServer())
          .get('/transactions')
          .expect(HttpStatus.OK);

        const { data } = body;
        expect((data as unknown[]).length).toBeGreaterThan(10);
        expect((data as unknown[])[0]).toMatchObject({
          id: expect.any(Number),
          hash: expect.any(String),
          fee: expect.any(String),
          size: expect.any(Number),
          timestamp: expect.any(String),
          block_id: block.id,
          notes: notes,
          spends: spends,
        });
        expect(((data as unknown[])[0] as Transaction).id).toBeGreaterThan(
          ((data as unknown[])[1] as Transaction).id,
        );
      });
    });
  });
});
