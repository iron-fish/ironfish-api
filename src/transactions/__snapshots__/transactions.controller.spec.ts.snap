// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionsController GET /transactions/find with an invalid hash returns a 404 1`] = `
Object {
  "message": "Not Found",
  "statusCode": 404,
}
`;

exports[`TransactionsController GET /transactions/find with an undefined hash returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "hash must be a string",
  ],
  "statusCode": 422,
}
`;

exports[`TransactionsController POST /transactions with a missing API key returns a 401 1`] = `
Object {
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`TransactionsController POST /transactions with missing arguments returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "transactions.0.hash must be a string",
    "transactions.0.fee must be an integer number",
    "transactions.0.size must be an integer number",
    "transactions.0.timestamp must be a Date instance",
    "transactions.0.block_id must be an integer number",
    "transactions.0.notes must be an array",
    "transactions.0.spends must be an array",
  ],
  "statusCode": 422,
}
`;

exports[`TransactionsController POST /transactions with too many transactions returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "transactions must contain not more than 3000 elements",
  ],
  "statusCode": 422,
}
`;
